FWCore/Framework/src/EventProcessor.cc

```diff
@@ -2131,25 +2155,53 @@ namespace edm {
 
   void EventProcessor::handleNextItemAfterMergingRunEntries(std::shared_ptr<RunProcessingStatus> iRunStatus,
                                                             WaitingTaskHolder iHolder) {
-    if (lastTransitionType() == InputSource::IsFile) {
-      iRunStatus->holderOfTaskInProcessRuns().doneWaiting(std::exception_ptr{});
-      iHolder.doneWaiting(std::exception_ptr{});
-    } else if (lastTransitionType() == InputSource::IsLumi && !iHolder.taskHasFailed()) {
-      CMS_SA_ALLOW try {
-        beginLumiAsync(IOVSyncValue(EventID(input_->run(), input_->luminosityBlock(), 0),
-                                    input_->luminosityBlockAuxiliary()->beginTime()),
-                       iRunStatus,
-                       iHolder);
-      } catch (...) {
-        WaitingTaskHolder copyHolder(iHolder);
-        iHolder.doneWaiting(std::current_exception());
-        endRunAsync(std::move(iRunStatus), std::move(iHolder));
+    chain::first([this, iRunStatus](auto nextTask) {
+      if (needToCallNext_) {
+        auto group = nextTask.group();
+        sourceResourcesAcquirer_.serialQueueChain().push(
+            *group, [this, iRunStatus, nextHolder = std::move(nextTask)]() mutable {
+              CMS_SA_ALLOW try {
+                ServiceRegistry::Operate operate(serviceToken_);
+                std::lock_guard<std::recursive_mutex> guard(*(sourceMutex_.get()));
+                nextTransitionType();
+                if (lastTransitionType() == InputSource::ItemType::IsRun &&
+                    iRunStatus->runPrincipal()->run() == input_->run() &&
+                    iRunStatus->runPrincipal()->reducedProcessHistoryID() == input_->reducedProcessHistoryID()) {
+                  throw Exception(errors::LogicError)
+                      << "InputSource claimed previous Run Entry was last to be merged in this file,\n"
+                      << "but the next entry has the same run number and reduced ProcessHistoryID.\n"
+                      << "This is probably a bug in the InputSource. Please report to the Core group.\n";
+                }
+              } catch (...) {
+                nextHolder.doneWaiting(std::current_exception());
+              }
+            });
       }
-    } else {
-      // Note that endRunAsync will call beginRunAsync for the following run
-      // if appropriate.
-      endRunAsync(std::move(iRunStatus), std::move(iHolder));
-    }
+    }) | chain::then([this, iRunStatus](std::exception_ptr const* iException, auto nextTask) {
```

Comment:
If you use `WaitingTaskHolder nextTask` then you have a copy and can call `doneWaiting` explicitly. Then in the `catch` clause you could do `nextTask.presetTaskAsFailed(std::current_exception())`. This would allow you to 
- add a `return` at line 2188,
-  replace the `else if` with `if`
-  remove the call to `endRunAsync` in the `catch` clause
- do away with the `else`

Reply:
Done. Thanks.